README
README
======

hutos
-----
hutos is Hurd toolchain building scripts. 

Project webpage is at:

http://savannah.nongnu.org/projects/hutos


Build
-----
Create a setenv symbolic link file that points to the gcc-*-glibc-*.sh
file before you want to build it. For example:

	$ ln -sf gcc-4.1-glibc-2.7.sh setenv

To build the toolchain run:

	$ sh hurd-toolchain.sh

If you would like to build offline, download the relevant SOURCES
present in the gcc-*-glibc-*.sh and cross-gnu, cross-gnu-env scripts
in .tar.bz2 file format and put it in a tarballs directory.

or

$ sh hurd-toolchain.sh archive
will create tarballs of already downloaded sources for later use

Cleanup
-------
$ sh hurd-toolchain.sh clean
will remove _all_ the downloaded sources+builds. If you don't want to download
the sources again use the "archive" argument before cleaning up.

References
----------
The scripts are based on the Makefile written from:
http://www.bddebian.com/~wiki/hurd/building/cross-compiling/Makefile.

The complete documentation is available at:
http://www.bddebian.com/~wiki/hurd/building/cross-compiling.

Contact
-------
Shakthi Kannan, shakthi.kannan@qvantel.com
Sudharshan "Sup3rkiddo" S, sudharsh@gmail.com

ChangeLog
2008-02-12 Sudharshan "Sup3rkiddo" S <sudharsh@gmail.com>
	* hurd-toolchain.sh: Added support for creating tarballs and cleaning up
	through command line arguments

	(create_tarball): New function
	
	
2008-02-08 Sudharshan "Sup3rkiddo" S <sudharsh@gmail.com>
	* HEAD: add gcc-4.2-glib-2.7 toolchain and related patches
	http://www.schwinge.homeip.net/~thomas/tmp/gcc-patches/

	* README: Removed information about copying patches as they are available
	in hutos CVS itself

	
2008-02-07 Sudharshan "Sup3rkiddo" S <sudharsh@gmail.com>
	* HEAD: added patches directory. Toolchain specific patches can now
	be included in the CVS repository itself
	
	* gcc-4.1-glibc-2.7.sh: Removed patches arrays PATCH0, PATCH1.
	Apply all patches present in patches dirs instead

	
2008-01-13 Sudharshan "Sup3rkiddo" S <sudharsh@gmail.com>
	* hurd-toolchain.sh: Add -p to mkdir added the missing 'mv src
	binutils'

	
2008-01-12 Shakthi Kannan <shakthi.kannan@qvantel.com>
	* hurd-toolchain.sh: Added build variables. 

	Removed apply_patches(). This needs to go into gcc-*-glibc-*.sh
	specific script.

	Using xjvf with tar. Using .tar.bz2 in ${TARBALLS} only.
	
	(get_cross_gnu): Added condition to use cross-gnu, cross-gnu-env
	if present in ${TARBALLS}

	(get_binutils): Remove old directory, if present.
	
	(get_gcc): Remove old directory, if present. Added
	apply_gcc_patches().

	(get_gnumach): Remove old directory, if present.

	(get_hurd): Remove old directory, if present.

	(get_glibc): Remove old directory, if present. Added
	apply_glibc_patches().

	* gcc-4.1-glibc-2.7.sh: Using binutils-2_17.

	Added TARBALL variables.

	(apply_glibc_patches): Updated apply_glibc_patches to use
	${TARBALLS}/${GLIBC}

	* README: Updated build instructions.


patch.sh
#!/bin/sh
#
# patch.sh 
#
# Copyright (C) 2007 Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any later
# version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# Written by Shakthi Kannan <shakthi.kannan@qvantel.com>.

get_apply_patches() {
    package_dir=$1   # $1 = package directory, ${GLIBC_DIR}
    patch_level=$2   # $2 = patch-level, p0/p1;
    patch_URL=$3     # $3 = patch directory/URL
    patches_array=$4 # $4 = list (array) of patches; 0 for all patches

    # TODO: do exception handling for above arguments?
    echo $1
    # create patch directory
    mkdir -p ${package_dir}/${patch_level}

    # starts with a . or /, means patches are in a directory
    first_char=`expr "$patch_URL" : '\(.\)'`

    # get patches
    case $first_char in
	.|/)if [ ! -z ${patches_array} ]; then # if 0, copy all patches
		for file in `ls ${patch_URL}/`; do
		    cp ${patch_URL}/$file ${package_dir}/${patch_level}
		done
	    fi ;;
	*)# default, a URL
	    if [ ! -z ${patches_array} ]; then # if 0, copy all patches
		    wget -r -np -nd -P ${package_dir}/${patch_level} ${patch_URL}/*
	    else # copy listed patches only
		index=0
		while [ $index -lt ${#patches_array[@]} ]; do
		    wget -r -np -nd -P ${package_dir}/${patch_level} ${patch_URL}/${patches_array[index]}
		    let "index = $index + 1"
		done
	    fi
    esac

    #apply patches
    index=0
    cd ${package_dir}

    if [ ! -z ${patches_array} ]; then # if 0, all patches to be applied
	for file in `ls ${patch_URL}/`; do
	    patch -${patch_level} -N < ${package_dir}/${patch_level}/${file}
	done
    else # only selected patches from array
	while [ $index -lt ${#patches_array[@]} ]; do
	    patch -${patch_level} -N < ${package_dir}/${patch_level}/${patches_array[index]}
	    let "index = $index + 1"
	done
    fi
}


hurd-toolchain.sh
#!/bin/sh
#
# hurd-toolchain.sh - Script for setting up Hurd toolchain builds
#
# Copyright (C) 2007 Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any later
# version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# Written by Shakthi Kannan <shakthi.kannan@qvantel.com>.

TOPDIR=`pwd`
DOWNLOADS=${TOPDIR}/downloads
ROOT=${TOPDIR}/root
PATCHES=${TOPDIR}/patches
SRC=${ROOT}/src
GLIBC_DIR=${SRC}/glibc
GCC_DIR=${SRC}/gcc
CROSS_GNU_URL=http://www.shakthimaan.com/downloads/hurd/toolchain
TARBALLS=${TOPDIR}/tarballs

all() {
    . ./patch.sh
    create_dir 
    get_sources 
    build_all
}

## Create directories
create_dir() {
	cd ${TOPDIR}
	mkdir -p ${DOWNLOADS}
	mkdir -p ${SRC}
}

get_sources() { 
    get_cross_gnu 
    get_binutils 
    get_gcc 
    get_gnumach 
    get_mig 
    get_hurd 
    get_glibc
}

get_cross_gnu() {
	echo "  ___ _ __ ___  ___ ___        __ _ _ __  _   _ "
	echo " / __| '__/ _ \/ __/ __|_____ / _\` | '_ \| | | |"
	echo "| (__| | | (_) \__ \__ \_____| (_| | | | | |_| |"
	echo " \___|_|  \___/|___/___/      \__, |_| |_|\__,_|"
	echo "                              |___/             "
	cd ${DOWNLOADS}
	if [ -f ${TARBALLS}/cross-gnu ] ; then
	    echo "Local copy of cross-gnu found..."
	    cp ${TARBALLS}/cross-gnu .
	else
		wget ${CROSS_GNU_URL}/cross-gnu -O cross-gnu
	fi
	echo "  ___ _ __ ___  ___ ___        __ _ _ __  _   _        ___ _ ____   __" 
	echo " / __| '__/ _ \/ __/ __|_____ / _\` | '_ \| | | |_____ / _ \ '_ \ \ / /"
	echo "| (__| | | (_) \__ \__ \_____| (_| | | | | |_| |_____|  __/ | | \ V / " 
	echo " \___|_|  \___/|___/___/      \__, |_| |_|\__,_|      \___|_| |_|\_/  " 
	echo "                              |___/                                   " 
	cd ${DOWNLOADS}
	if [ -f ${TARBALLS}/cross-gnu-env ] ; then
	    echo "Local copy of cross-gnu-env found..."
	    cp ${TARBALLS}/cross-gnu-env .
	else
	    wget ${CROSS_GNU_URL}/cross-gnu-env -O cross-gnu-env
	fi
	chmod +x cross-gnu
	chmod +x cross-gnu-env
}

get_binutils() {
	echo " _     _             _   _ _     "
	echo "| |__ (_)_ __  _   _| |_(_) |___ "
	echo "| '_ \| | '_ \| | | | __| | / __|"
	echo "| |_) | | | | | |_| | |_| | \__ \\"
	echo "|_.__/|_|_| |_|\__,_|\__|_|_|___/"
	cd ${SRC} 
	if test -d binutils; then
	    echo "Deleting existing sources for sanity. If you don't want this behaviour, "
	    echo "consider creating a tarball of the sources. Refer README for more"
	    echo "...."
	    rm -rf binutils
	fi
	if [ -f ${BINUTILS_TARBALL} ] ; then
	    echo "Local copy of BINUTILS tarball found. Extracting..."
	    tar xjvf ${BINUTILS_TARBALL}
	else
	    cvs -d:pserver:anoncvs@sources.redhat.com:/cvs/src co -r ${BINUTILS} binutils
	    mv src binutils
	fi
}

get_gcc() {
	echo "  __ _  ___ ___ "
	echo " / _\` |/ __/ __|"
	echo "| (_| | (_| (__ "
	echo " \__, |\___\___|"
	echo " |___/          "
	cd ${SRC}
    	if test -d gcc; then
	    echo "Deleting existing sources for sanity. If you don't want this behaviour, "
	    echo "consider creating a tarball of the sources. Refer README for more"
	    echo "...."	   
	    rm -rf gcc
	fi
	if [ -f ${GCC_TARBALL} ] ; then
	    echo "Local copy of gcc tarball found. Extracting..."
	    tar xjvf ${GCC_TARBALL}
	else
	    svn co svn://gcc.gnu.org/svn/gcc/branches/${GCC}
	    mv ${GCC} gcc
	fi
	apply_gcc_patches
	cd gcc/ && contrib/gcc_update --touch
}

get_gnumach() {
	echo "                                        _     "
	echo "  __ _ _ __  _   _ _ __ ___   __ _  ___| |__  "
	echo " / _\` | '_ \| | | | '_ \` _ \ / _\` |/ __| '_ \ "
	echo "| (_| | | | | |_| | | | | | | (_| | (__| | | |"
	echo " \__, |_| |_|\__,_|_| |_| |_|\__,_|\___|_| |_|"
	echo " |___/                                        "
	cd ${SRC}
	if test -d gnumach; then
	    echo "Deleting existing sources for sanity. If you don't want this behaviour, "
	    echo "consider creating a tarball of the sources. Refer README for more"
	    echo "...."	   
	    rm -rf gnumach
	fi
	if [ -f ${GNUMACH_TARBALL} ] ; then
	    echo "Local copy of GNU MACH tarball found. Extracting..."
	    tar xjvf ${GNUMACH_TARBALL}
	else
	    cvs -d:pserver:anoncvs@cvs.gnu.org:/cvsroot/hurd co -r ${GNUMACH} gnumach
	fi
	cd gnumach/ && autoreconf -vfi
}

get_mig() {
	echo "           _       "
	echo " _ __ ___ (_) __ _ "
	echo "| '_ \` _ \| |/ _\` |"
	echo "| | | | | | | (_| |"
	echo "|_| |_| |_|_|\__, |"
	echo "             |___/ "
	cd ${SRC}
	if test -d mig; then
	    echo "Deleting existing sources for sanity. If you don't want this behaviour, "
	    echo "consider creating a tarball of the sources. Refer README for more"
	    echo "...."	   
	    rm -rf mig
	fi
	if [ -f ${MIG_TARBALL} ] ; then
	    echo "Local copy of MIG tarball found. Extracting..."
	    tar xjvf ${MIG_TARBALL}
	else
	    cvs -d:pserver:anoncvs@cvs.gnu.org:/cvsroot/hurd co ${MIG}
	fi
	cd mig/ && autoreconf -vfi
}

get_hurd() {
	echo " _                   _ "
	echo "| |__  _   _ _ __ __| |"
	echo "| '_ \| | | | '__/ _\` |"
	echo "| | | | |_| | | | (_| |"
	echo "|_| |_|\__,_|_|  \__,_|"
	cd ${SRC}
	if test -d hurd; then
	    echo "Deleting existing sources for sanity. If you don't want this behaviour, "
	    echo "consider creating a tarball of the sources. Refer README for more"
	    echo "...."	   
	    rm -rf hurd
	fi
	if [ -f ${HURD_TARBALL} ] ; then
	    echo "Local copy of HURD tarball found. Extracting..."
	    tar xjvf ${HURD_TARBALL}
	else
	    cvs -d:pserver:anoncvs@cvs.gnu.org:/cvsroot/hurd co ${HURD}
	fi
}

get_glibc() {
	echo "       _ _ _          "
	echo "  __ _| (_) |__   ___ "
	echo " / _\` | | | '_ \ / __|"
	echo "| (_| | | | |_) | (__ "
	echo " \__, |_|_|_.__/ \___|"
	echo " |___/                "
	cd ${SRC}
	if test -d glibc; then
	    echo "Deleting existing sources for sanity. If you don't want this behaviour, "
	    echo "consider creating a tarball of the sources. Refer README for more"
	    echo "...."	   
	    rm -rf glibc
	fi
	if [ -f ${GLIBC_TARBALL} ] ; then
	    echo "Local copy of glibc tarball found. Extracting..."
	    tar xjvf ${GLIBC_TARBALL}
	else
	    cvs -d:pserver:anoncvs@sources.redhat.com:/cvs/glibc co -r ${GLIBC} glibc
	    mv libc glibc
	fi
	apply_glibc_patches
}

build_all() {
	ROOT=${TOPDIR}/root
	export PATH="$PATH:${DOWNLOADS}"
	. ${DOWNLOADS}/cross-gnu-env
	${DOWNLOADS}/cross-gnu
}

clean() {
	rm -rf downloads root *~
}

create_tarball() {
        echo "Creating tarball for $1"
        if ! [ -d ${TARBALLS} ]
	then
	    mkdir ${TARBALLS}
	fi
	cd root/src
	tar -cjvf ${TARBALLS}/$1.tar.bz2 $2
}	    

## Execute
. ./setenv
if [ "$1" = "clean" ]
then
    echo "Cleaning builds/sources"
    rm -rf root
elif [ "$1" = "archive" ]
    then
    echo "Archiving sources..."
    cd root/src
    create_tarball ${GCC} gcc
    create_tarball ${GLIBC} glibc
    create_tarball ${BINUTILS} binutils
    create_tarball ${HURD} hurd
    create_tarball ${GNUMACH} gnumach
    create_tarball ${MIG} mig
    echo "Done creating Tarballs"
elif [ "$1" = "help" ]
    then
    echo "hurd-toolchain, wrapper script for crossgnu to create a cross environment for GNU/Hurd"
    echo "Available options"
    echo "archive : Create an archive of the sources for later offline build"
    echo "clean   : Cleanup sources and objects"
else
    all
fi


gcc-4.1-glibc-2.7.sh
#!/bin/sh
#
# gcc-4.1-glibc-2.7.sh 
#
# Copyright (C) 2007 Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any later
# version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# Written by Shakthi Kannan <shakthi.kannan@qvantel.com>.

## SOURCES
BINUTILS=binutils-2_17-branch
GCC=gcc-4_1-branch
GNUMACH=gnumach-1-branch
GLIBC=glibc-2_7-branch
MIG=mig
HURD=hurd

GCC_TARBALL=${TARBALLS}/${GCC}.tar.bz2
BINUTILS_TARBALL=${TARBALLS}/${BINUTILS}.tar.bz2
GNUMACH_TARBALL=${TARBALLS}/${GNUMACH}.tar.bz2
GLIBC_TARBALL=${TARBALLS}/${GLIBC}.tar.bz2
MIG_TARBALL=${TARBALLS}/${MIG}.tar.bz2
HURD_TARBALL=${TARBALLS}/${HURD}.tar.bz2

# $1 = package directory, ${GLIBC_DIR}
# $2 = patch-level, p0/p1;
# $3 = patch directory/URL
# $4 = list (array) of patches; 0 for all patches

apply_glibc_patches() {
    get_apply_patches ${GLIBC_DIR} p0 "${PATCHES}/${GLIBC}/p0" 0
    get_apply_patches ${GLIBC_DIR} p1 "${PATCHES}/${GLIBC}/p1" 0 
}

#apply_gcc_patches() {
## Nothing for gcc-4.1
#}

gcc-4.2-glibc-2.7.sh
#!/bin/sh
#
# gcc-4.2-glibc-2.7.sh 
#
# Copyright (C) 2007 Free Software Foundation, Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any later
# version.
# 
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
# 
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
#
# Written by Shakthi Kannan <shakthi.kannan@qvantel.com>
# and Sudharshan "Sup3rkiddo" S <sudharsh@gmail.com>.

## SOURCES
BINUTILS=binutils-2_17-branch
GCC=gcc-4_2-branch
GNUMACH=gnumach-1-branch
GLIBC=glibc-2_7-branch
MIG=mig
HURD=hurd

GCC_TARBALL=${TARBALLS}/${GCC}.tar.bz2
BINUTILS_TARBALL=${TARBALLS}/${BINUTILS}.tar.bz2
GNUMACH_TARBALL=${TARBALLS}/${GNUMACH}.tar.bz2
GLIBC_TARBALL=${TARBALLS}/${GLIBC}.tar.bz2
MIG_TARBALL=${TARBALLS}/${MIG}.tar.bz2
HURD_TARBALL=${TARBALLS}/${HURD}.tar.bz2

# $1 = package directory, ${GLIBC_DIR}
# $2 = patch-level, p0/p1;
# $3 = patch directory/URL
# $4 = list (array) of patches; 0 for all patches

apply_glibc_patches() {
    get_apply_patches ${GLIBC_DIR} p0 "${PATCHES}/${GLIBC}/p0" 0
    get_apply_patches ${GLIBC_DIR} p1 "${PATCHES}/${GLIBC}/p1" 0 
}

apply_gcc_patches() {
    get_apply_patches ${GCC_DIR} p0 "${PATCHES}/${GCC}/p0" 0
}